In [7]:
import pandas as pd
from scipy.stats import skew, kurtosis

data = pd.read_csv('electricity_data.csv')

In [2]:
data

record_id  utility_id_ferc1  report_year  \
0        f1_fuel_1994_12_1_0_7                 1         1994   
1       f1_fuel_1994_12_1_0_10                 1         1994   
2        f1_fuel_1994_12_2_0_1                 2         1994   
3        f1_fuel_1994_12_2_0_7                 2         1994   
4       f1_fuel_1994_12_2_0_10                 2         1994   
...                        ...               ...          ...   
29518  f1_fuel_2018_12_12_0_13                12         2018   
29519   f1_fuel_2018_12_12_1_1                12         2018   
29520  f1_fuel_2018_12_12_1_10                12         2018   
29521  f1_fuel_2018_12_12_1_13                12         2018   
29522  f1_fuel_2018_12_12_1_14                12         2018   

           plant_name_ferc1 fuel_type_code_pudl fuel_unit  fuel_qty_burned  \
0                  rockport                coal       ton        5377489.0   
1      rockport total plant                coal       ton       10486945.0   
2                    gorgas                coal       ton        2978683.0   
3                     barry                coal       ton        3739484.0   
4                 chickasaw                 gas       mcf          40533.0   
...                     ...                 ...       ...              ...   
29518    neil simpson ct #1                 gas       mcf          18799.0   
29519  cheyenne prairie 58%                 gas       mcf         806730.0   
29520     lange ct facility                 gas       mcf         104554.0   
29521       wygen 3 bhp 52%                coal       ton         315945.0   
29522       wygen 3 bhp 52%                 gas       mcf          17853.0   

       fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
0                   16.590                      18.59   
1                   16.592                      18.58   
2                   24.130                      39.72   
3                   23.950                      47.21   
4                    1.000                       2.77   
...                    ...                        ...   
29518                1.059                       4.78   
29519                1.050                       3.65   
29520                1.060                       4.77   
29521               16.108                       3.06   
29522                1.059                       0.00   

       fuel_cost_per_unit_delivered  fuel_cost_per_mmbtu  
0                             18.53                1.121  
1                             18.53                1.120  
2                             38.12                1.650  
3                             45.99                1.970  
4                              2.77                2.570  
...                             ...                  ...  
29518                          4.78                9.030  
29519                          3.65                6.950  
29520                          4.77                8.990  
29521                         14.76                1.110  
29522                          0.00               11.680  

[29523 rows x 11 columns]

In [4]:
data.describe()[['fuel_mmbtu_per_unit']].loc[['std', '75%'], 'fuel_mmbtu_per_unit'].reset_index()

index  fuel_mmbtu_per_unit
0   std             10.60022
1   75%             17.00600

#### The std and 75% percentile is 10.60 and 17.01 rspectively

In [6]:
data.groupby('fuel_type_code_pudl')['fuel_cost_per_unit_burned'].mean().reset_index().sort_values('fuel_cost_per_unit_burned')

fuel_type_code_pudl  fuel_cost_per_unit_burned
1                 gas                  13.659397
4               other                  18.253856
5               waste                  19.518122
0                coal                  67.421830
3                 oil                 168.877086
2             nuclear                4955.157002

#### From the code, the fuel type code with the lowest average fuel cost per unit burned is "gas"

In [8]:
print(skew(data['fuel_qty_burned']))
print()
print(kurtosis(data['fuel_qty_burned']))

15.850690077422112

651.2589365474387


#### The skew and kurtosis values are 15.85 and 651.26 rsepctively

In [9]:
data.isnull().sum()

record_id                         0
utility_id_ferc1                  0
report_year                       0
plant_name_ferc1                  0
fuel_type_code_pudl               0
fuel_unit                       180
fuel_qty_burned                   0
fuel_mmbtu_per_unit               0
fuel_cost_per_unit_burned         0
fuel_cost_per_unit_delivered      0
fuel_cost_per_mmbtu               0
dtype: int64

In [11]:
(data.isnull().sum()['fuel_unit']/len(data)) * 100

0.609694136774718

#### The column with missing values is "fuel_unit", while the number of missing values are "180" and the percentage is "0.610"

In [12]:
data.corr()['fuel_cost_per_unit_burned'].reset_index().sort_values('fuel_cost_per_unit_burned')

index  fuel_cost_per_unit_burned
0              utility_id_ferc1                  -0.037863
2               fuel_qty_burned                  -0.018535
3           fuel_mmbtu_per_unit                  -0.010034
6           fuel_cost_per_mmbtu                  -0.000437
5  fuel_cost_per_unit_delivered                   0.011007
1                   report_year                   0.013599
4     fuel_cost_per_unit_burned                   1.000000

#### The feature with the second and third lowest correlation with fuel_cost_per_unit_burned is "fuel_qty_burned" and "fuel_mmbtu_per_unit" respectively

In [14]:
year_groupby = data.groupby('report_year')['fuel_cost_per_unit_delivered'].mean().reset_index()
year_groupby.sort_values('fuel_cost_per_unit_delivered', ascending = False)

report_year  fuel_cost_per_unit_delivered
3          1997                  11140.197239
2          1996                   9196.705948
6          2000                    985.362877
15         2009                    652.694163
24         2018                    499.269966
21         2015                    326.535511
4          1998                    287.154420
20         2014                    192.737183
19         2013                    172.307591
10         2004                    139.524275
22         2016                    103.901761
16         2010                     91.862105
0          1994                     63.636060
18         2012                     60.994502
7          2001                     60.050396
17         2011                     59.774667
14         2008                     58.588197
9          2003                     55.663493
8          2002                     47.594361
23         2017                     46.196861
13         2007                     43.325023
11         2005                     41.438184
12         2006                     38.657484
1          1995                     32.735269
5          1999                     25.551627

#### From the result above, the year with the highest average fuel cost per unit learnt is "1997"

In [22]:
fuel_type = data.groupby(['fuel_type_code_pudl','report_year'])['fuel_cost_per_unit_burned'].sum().reset_index()
fuel_type = fuel_type.set_index('report_year')
fuel_type = fuel_type[fuel_type['fuel_type_code_pudl'] == 'coal']
fuel_type

fuel_type_code_pudl  fuel_cost_per_unit_burned
report_year                                               
1994                       coal                  14984.572
1995                       coal                  14571.785
1996                       coal                  12694.803
1997                       coal                  43742.178
1998                       coal                  11902.597
1999                       coal                  16303.468
2000                       coal                 100428.831
2001                       coal                  15109.012
2002                       coal                   9171.164
2003                       coal                  30954.110
2004                       coal                   9632.063
2005                       coal                  11107.182
2006                       coal                  12402.684
2007                       coal                  12965.171
2008                       coal                  14331.348
2009                       coal                  45554.858
2010                       coal                  88627.745
2011                       coal                  17001.015
2012                       coal                  17546.097
2013                       coal                  17629.860
2014                       coal                  15412.197
2015                       coal                  13338.235
2016                       coal                  11100.925
2017                       coal                  10254.284
2018                       coal                   9488.197

In [25]:
year_1994 = fuel_type['fuel_cost_per_unit_burned'][1994] 
year_1998 = fuel_type['fuel_cost_per_unit_burned'][1998]

((year_1994 - year_1998)/year_1994) * 100

20.56765451826049

#### The percentage change in fuel cost per unit burned in 1998 compared to 1994 for coal fuel type is......"21%"